frontendUiTableImpl/src/main/java/com/virtuslab/gitmachete/frontend/ui/impl/table/EnhancedGraphTable.java

```diff
@@ -218,51 +216,49 @@ private void doAutomaticDiscover(Path macheteFilePath) {
     var mainDirPath = GitVfsUtils.getMainDirectoryPath(selectedRepository).toAbsolutePath();
     var gitDirPath = GitVfsUtils.getGitDirectoryPath(selectedRepository).toAbsolutePath();
 
-    @SuppressWarnings({"nullness", "guieffect"}) // To allow runResult be uninitialized and Tuple accept @Nullable IBranchLayout
-    var task = new Task.Backgroundable(project,
+    // `nullness` to allow runResult be uninitialized and IBranchLayoutWriter#write accept @Nullable IBranchLayout without redundant check
+    // `guieffect` to allow running lambdas that are @AlwaysSafe in methods onSuccess and onFailure
+    @SuppressWarnings({"nullness", "guieffect"}) var task = new Task.Backgroundable(project,
         getString("string.GitMachete.EnhancedGraphTable.automatic-discover.task-title")) {
-      private Try<Tuple3<IBranchLayoutWriter, @Nullable IBranchLayout, IGitMacheteRepositorySnapshot>> runResult;
+      private Try<DiscoverResult> discoverResult;
 
       @Override
       public void run(ProgressIndicator indicator) {
-        runResult = Try.of(() -> RuntimeBinding.instantiateSoleImplementingClass(IGitMacheteRepositoryCache.class)
+        discoverResult = Try.of(() -> RuntimeBinding.instantiateSoleImplementingClass(IGitMacheteRepositoryCache.class)
             .getInstance(mainDirPath, gitDirPath).discoverLayoutAndCreateSnapshot())
             .filterTry(repositorySnapshot -> repositorySnapshot.getRootBranches().size() > 0,
```

Comment:
Hmm how about the following: in case the condition is not satisfied, just `GuiUtils.invokeLater` the `setTextForEmptyTable(...)` piece and return from the method? Similarly for the case where `result.branchLayout == null` below. This way, the intermediate exceptions won't be needed and the entire flow will be likely easier to analyze

Reply:
revamped